TestAccAwsRamResourceShareAccepter_basic fails

<!---
Please note the following potential times when an issue might be in Terraform core:

* [Configuration Language](https://www.terraform.io/docs/configuration/index.html) or resource ordering issues
* [State](https://www.terraform.io/docs/state/index.html) and [State Backend](https://www.terraform.io/docs/backends/index.html) issues
* [Provisioner](https://www.terraform.io/docs/provisioners/index.html) issues
* [Registry](https://registry.terraform.io/) issues
* Spans resources across multiple providers

If you are running into one of these scenarios, we recommend opening an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/) instead.
--->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

<!--- Please run `terraform -v` to show the Terraform core version and provider version(s). If you are not running the latest version of Terraform or the provider, please upgrade because your issue may have already been fixed. [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions). --->

```
Terraform v0.13.5
```

And terraform-provider-aws is the latest master.

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->

* aws_ram_resource_share_accepter

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

See [testAccAwsRamResourceShareAccepterBasic](https://github.com/hashicorp/terraform-provider-aws/blob/81da040d9fd06811b7e007533c829ec72f299543/aws/resource_aws_ram_resource_share_accepter_test.go#L108).

### Expected Behavior

The test passes

### Actual Behavior

The test fails with the following message:

```
$ make testacc TESTARGS='-run=TestAccAwsRamResourceShareAccepter_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsRamResourceShareAccepter_basic -timeout 120m
=== RUN   TestAccAwsRamResourceShareAccepter_basic
=== PAUSE TestAccAwsRamResourceShareAccepter_basic
=== CONT  TestAccAwsRamResourceShareAccepter_basic
    TestAccAwsRamResourceShareAccepter_basic: resource_aws_ram_resource_share_accepter_test.go:23: Step 1/2 error: Error running apply:
        Error: No RAM Resource Share (arn:aws:ram:ap-northeast-1:732481906906:resource-share/f65079e2-e952-4014-8a54-500b19ea767a) invitation found

        NOTE: If both AWS accounts are in the same AWS Organization and RAM Sharing with AWS Organizations is enabled, this resource is not necessary


--- FAIL: TestAccAwsRamResourceShareAccepter_basic (24.27s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	26.148s
FAIL
make: *** [testacc] Error 1
```

```
make testacc TESTARGS='-run=TestAccAwsRamResourceShareAccepter_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsRamResourceShareAccepter_basic -timeout 120m
=== RUN   TestAccAwsRamResourceShareAccepter_basic
=== PAUSE TestAccAwsRamResourceShareAccepter_basic
=== CONT  TestAccAwsRamResourceShareAccepter_basic
    resource_aws_ram_resource_share_accepter_test.go:23: Step 1/2 error: Error running apply:
        Error: error retrieving resource shares: UnknownResourceException: ResourceShare arn:aws:ram:ap-northeast-1:732481906906:resource-share/b57bc5aa-7736-4cae-babe-c69e9c2592ed could not be found.


--- FAIL: TestAccAwsRamResourceShareAccepter_basic (28.01s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	29.992s
FAIL
make: *** [testacc] Error 1
```
### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor documentation? For example:
--->

* https://github.com/hashicorp/terraform-provider-aws/issues/13494

When I implementing https://github.com/hashicorp/terraform-provider-aws/pull/16576, I noticed `aws_ram_resource_share_accepter` kept failing, so I ran its ACC test and the test failed too with the same error. I'm going to look into the problem later but if someone has any ideas about what is wrong with it, I'd like to know.

--- a/.changelog/17032.txt
+++ b/.changelog/17032.txt
@@ -1,0 +1,7 @@
+```release-note:bug
+resource/aws_ram_resource_share_accepter: Improve handling of eventual consistency
+```
+
+```release-note:bug
+resource/aws_ram_resource_share: Improve handling of eventual consistency
+```

--- a/.changelog/17032.txt
+++ b/.changelog/17032.txt
@@ -5,3 +5,7 @@ resource/aws_ram_resource_share_accepter: Improve handling of eventual consisten
 ```release-note:bug
 resource/aws_ram_resource_share: Improve handling of eventual consistency
 ```
+
+```release-note:bug
+resource/aws_ram_principal_association: Improve handling of eventual consistency
+```

--- a/aws/internal/service/ram/finder/finder.go
+++ b/aws/internal/service/ram/finder/finder.go
@@ -10,6 +10,11 @@ import (
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
+const (
+	FindInvitationTimeout    = 2 * time.Minute
+	FindResourceShareTimeout = 1 * time.Minute
+)
+
 // ResourceShareOwnerOtherAccountsByArn returns the resource share owned by other accounts corresponding to the specified ARN.
 // Returns nil if no configuration is found.
 func ResourceShareOwnerOtherAccountsByArn(conn *ram.RAM, arn string) (*ram.ResourceShare, error) {
@@ -38,7 +43,7 @@ func ResourceShareInvitationByResourceShareArnAndStatus(conn *ram.RAM, resourceS
 	var invitation *ram.ResourceShareInvitation
 
 	// Retry for Ram resource share invitation eventual consistency
-	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+	err := resource.Retry(FindInvitationTimeout, func() *resource.RetryError {
 		i, err := resourceShareInvitationByResourceShareArnAndStatus(conn, resourceShareArn, status)
 		invitation = i
 
@@ -74,7 +79,7 @@ func ResourceShareInvitationByArn(conn *ram.RAM, arn string) (*ram.ResourceShare
 	var invitation *ram.ResourceShareInvitation
 
 	// Retry for Ram resource share invitation eventual consistency
-	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+	err := resource.Retry(FindInvitationTimeout, func() *resource.RetryError {
 		i, err := resourceShareInvitationByArn(conn, arn)
 		invitation = i
 
@@ -108,7 +113,7 @@ func resourceShare(conn *ram.RAM, input *ram.GetResourceSharesInput) (*ram.Resou
 	var shares *ram.GetResourceSharesOutput
 
 	// Retry for Ram resource share eventual consistency
-	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+	err := resource.Retry(FindResourceShareTimeout, func() *resource.RetryError {
 		ss, err := conn.GetResourceShares(input)
 		shares = ss
 
@@ -183,4 +188,24 @@ func resourceShareInvitationByArn(conn *ram.RAM, arn string) (*ram.ResourceShare
 	}
 
 	return output.ResourceShareInvitations[0], nil
+}
+
+func ResourceSharePrincipalAssociationByShareARNPrincipal(conn *ram.RAM, resourceShareARN, principal string) (*ram.ResourceShareAssociation, error) {
+	input := &ram.GetResourceShareAssociationsInput{
+		AssociationType:   aws.String(ram.ResourceShareAssociationTypePrincipal),
+		Principal:         aws.String(principal),
+		ResourceShareArns: aws.StringSlice([]string{resourceShareARN}),
+	}
+
+	output, err := conn.GetResourceShareAssociations(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if output == nil || len(output.ResourceShareAssociations) == 0 || output.ResourceShareAssociations[0] == nil {
+		return nil, nil
+	}
+
+	return output.ResourceShareAssociations[0], nil
 }

--- a/aws/internal/service/ram/finder/finder.go
+++ b/aws/internal/service/ram/finder/finder.go
@@ -1,0 +1,186 @@
+package finder
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ram"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
+)
+
+// ResourceShareOwnerOtherAccountsByArn returns the resource share owned by other accounts corresponding to the specified ARN.
+// Returns nil if no configuration is found.
+func ResourceShareOwnerOtherAccountsByArn(conn *ram.RAM, arn string) (*ram.ResourceShare, error) {
+	listResourceSharesInput := &ram.GetResourceSharesInput{
+		ResourceOwner:     aws.String(ram.ResourceOwnerOtherAccounts),
+		ResourceShareArns: aws.StringSlice([]string{arn}),
+	}
+
+	return resourceShare(conn, listResourceSharesInput)
+}
+
+// ResourceShareOwnerSelfByArn returns the resource share owned by own account corresponding to the specified ARN.
+// Returns nil if no configuration is found.
+func ResourceShareOwnerSelfByArn(conn *ram.RAM, arn string) (*ram.ResourceShare, error) {
+	listResourceSharesInput := &ram.GetResourceSharesInput{
+		ResourceOwner:     aws.String(ram.ResourceOwnerSelf),
+		ResourceShareArns: aws.StringSlice([]string{arn}),
+	}
+
+	return resourceShare(conn, listResourceSharesInput)
+}
+
+// ResourceShareInvitationByResourceShareArnAndStatus returns the resource share invitation corresponding to the specified resource share ARN.
+// Returns nil if no configuration is found.
+func ResourceShareInvitationByResourceShareArnAndStatus(conn *ram.RAM, resourceShareArn, status string) (*ram.ResourceShareInvitation, error) {
+	var invitation *ram.ResourceShareInvitation
+
+	// Retry for Ram resource share invitation eventual consistency
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		i, err := resourceShareInvitationByResourceShareArnAndStatus(conn, resourceShareArn, status)
+		invitation = i
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		if invitation == nil {
+			return resource.RetryableError(&resource.NotFoundError{})
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		invitation, err = resourceShareInvitationByResourceShareArnAndStatus(conn, resourceShareArn, status)
+	}
+
+	if invitation == nil {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return invitation, nil
+}
+
+// ResourceShareInvitationByArn returns the resource share invitation corresponding to the specified ARN.
+// Returns nil if no configuration is found.
+func ResourceShareInvitationByArn(conn *ram.RAM, arn string) (*ram.ResourceShareInvitation, error) {
+	var invitation *ram.ResourceShareInvitation
+
+	// Retry for Ram resource share invitation eventual consistency
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		i, err := resourceShareInvitationByArn(conn, arn)
+		invitation = i
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		if invitation == nil {
+			resource.RetryableError(&resource.NotFoundError{})
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		invitation, err = resourceShareInvitationByArn(conn, arn)
+	}
+
+	if invitation == nil {
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return invitation, nil
+}
+
+func resourceShare(conn *ram.RAM, input *ram.GetResourceSharesInput) (*ram.ResourceShare, error) {
+	var shares *ram.GetResourceSharesOutput
+
+	// Retry for Ram resource share eventual consistency
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		ss, err := conn.GetResourceShares(input)
+		shares = ss
+
+		if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
+			return resource.RetryableError(err)
+		}
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		if len(shares.ResourceShares) == 0 {
+			return resource.RetryableError(&resource.NotFoundError{})
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		shares, err = conn.GetResourceShares(input)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if shares == nil || len(shares.ResourceShares) == 0 {
+		return nil, nil
+	}
+
+	return shares.ResourceShares[0], nil
+}
+
+func resourceShareInvitationByResourceShareArnAndStatus(conn *ram.RAM, resourceShareArn, status string) (*ram.ResourceShareInvitation, error) {
+	var invitation *ram.ResourceShareInvitation
+
+	input := &ram.GetResourceShareInvitationsInput{
+		ResourceShareArns: []*string{aws.String(resourceShareArn)},
+	}
+
+	err := conn.GetResourceShareInvitationsPages(input, func(page *ram.GetResourceShareInvitationsOutput, lastPage bool) bool {
+		for _, rsi := range page.ResourceShareInvitations {
+			if aws.StringValue(rsi.Status) == status {
+				invitation = rsi
+				return false
+			}
+		}
+
+		return !lastPage
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return invitation, nil
+}
+
+func resourceShareInvitationByArn(conn *ram.RAM, arn string) (*ram.ResourceShareInvitation, error) {
+	input := &ram.GetResourceShareInvitationsInput{
+		ResourceShareInvitationArns: []*string{aws.String(arn)},
+	}
+
+	output, err := conn.GetResourceShareInvitations(input)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(output.ResourceShareInvitations) == 0 {
+		return nil, nil
+	}
+
+	return output.ResourceShareInvitations[0], nil
+}

--- a/aws/internal/service/ram/waiter/status.go
+++ b/aws/internal/service/ram/waiter/status.go
@@ -1,0 +1,55 @@
+package waiter
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ram"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/finder"
+)
+
+const (
+	resourceShareInvitationStatusNotFound = "NotFound"
+	resourceShareInvitationStatusUnknown  = "Unknown"
+
+	resourceShareStatusNotFound = "NotFound"
+	resourceShareStatusUnknown  = "Unknown"
+)
+
+// ResourceShareInvitationStatus fetches the ResourceShareInvitation and its Status
+func ResourceShareInvitationStatus(conn *ram.RAM, arn string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		invitation, err := finder.ResourceShareInvitationByArn(conn, arn)
+
+		if err != nil {
+			return nil, resourceShareInvitationStatusUnknown, err
+		}
+
+		if invitation == nil {
+			return nil, resourceShareInvitationStatusNotFound, nil
+		}
+
+		return invitation, aws.StringValue(invitation.Status), nil
+	}
+}
+
+// ResourceShareOwnerSelfStatus fetches the ResourceShare and its Status
+func ResourceShareOwnerSelfStatus(conn *ram.RAM, arn string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		share, err := finder.ResourceShareOwnerSelfByArn(conn, arn)
+
+		if err != nil {
+			if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
+				return nil, resourceShareStatusNotFound, nil
+			}
+
+			return nil, resourceShareStatusUnknown, err
+		}
+
+		if share == nil {
+			return nil, resourceShareStatusNotFound, nil
+		}
+
+		return share, aws.StringValue(share.Status), nil
+	}
+}

--- a/aws/internal/service/ram/waiter/status.go
+++ b/aws/internal/service/ram/waiter/status.go
@@ -1,6 +1,8 @@
 package waiter
 
 import (
+	"fmt"
+
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ram"
 	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
@@ -9,11 +11,13 @@ import (
 )
 
 const (
-	resourceShareInvitationStatusNotFound = "NotFound"
-	resourceShareInvitationStatusUnknown  = "Unknown"
+	ResourceShareInvitationStatusNotFound = "NotFound"
+	ResourceShareInvitationStatusUnknown  = "Unknown"
 
-	resourceShareStatusNotFound = "NotFound"
-	resourceShareStatusUnknown  = "Unknown"
+	ResourceShareStatusNotFound = "NotFound"
+	ResourceShareStatusUnknown  = "Unknown"
+
+	PrincipalAssociationStatusNotFound = "NotFound"
 )
 
 // ResourceShareInvitationStatus fetches the ResourceShareInvitation and its Status
@@ -22,11 +26,11 @@ func ResourceShareInvitationStatus(conn *ram.RAM, arn string) resource.StateRefr
 		invitation, err := finder.ResourceShareInvitationByArn(conn, arn)
 
 		if err != nil {
-			return nil, resourceShareInvitationStatusUnknown, err
+			return nil, ResourceShareInvitationStatusUnknown, err
 		}
 
 		if invitation == nil {
-			return nil, resourceShareInvitationStatusNotFound, nil
+			return nil, ResourceShareInvitationStatusNotFound, nil
 		}
 
 		return invitation, aws.StringValue(invitation.Status), nil
@@ -38,18 +42,43 @@ func ResourceShareOwnerSelfStatus(conn *ram.RAM, arn string) resource.StateRefre
 	return func() (interface{}, string, error) {
 		share, err := finder.ResourceShareOwnerSelfByArn(conn, arn)
 
-		if err != nil {
-			if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
-				return nil, resourceShareStatusNotFound, nil
-			}
+		if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
+			return nil, ResourceShareStatusNotFound, nil
+		}
 
-			return nil, resourceShareStatusUnknown, err
+		if err != nil {
+			return nil, ResourceShareStatusUnknown, err
 		}
 
 		if share == nil {
-			return nil, resourceShareStatusNotFound, nil
+			return nil, ResourceShareStatusNotFound, nil
 		}
 
 		return share, aws.StringValue(share.Status), nil
+	}
+}
+
+func ResourceSharePrincipalAssociationStatus(conn *ram.RAM, resourceShareArn, principal string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		association, err := finder.ResourceSharePrincipalAssociationByShareARNPrincipal(conn, resourceShareArn, principal)
+
+		if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
+			return nil, PrincipalAssociationStatusNotFound, err
+		}
+
+		if err != nil {
+			return nil, ram.ResourceShareAssociationStatusFailed, err
+		}
+
+		if association == nil {
+			return nil, ram.ResourceShareAssociationStatusDisassociated, nil
+		}
+
+		if aws.StringValue(association.Status) == ram.ResourceShareAssociationStatusFailed {
+			extendedErr := fmt.Errorf("association status message: %s", aws.StringValue(association.StatusMessage))
+			return association, aws.StringValue(association.Status), extendedErr
+		}
+
+		return association, aws.StringValue(association.Status), nil
 	}
 }

--- a/aws/internal/service/ram/waiter/waiter.go
+++ b/aws/internal/service/ram/waiter/waiter.go
@@ -1,0 +1,80 @@
+package waiter
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/ram"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+// ResourceShareInvitationAccepted waits for a ResourceShareInvitation to return ACCEPTED
+func ResourceShareInvitationAccepted(conn *ram.RAM, arn string, timeout time.Duration) (*ram.ResourceShareInvitation, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareInvitationStatusPending},
+		Target:  []string{ram.ResourceShareInvitationStatusAccepted},
+		Refresh: ResourceShareInvitationStatus(conn, arn),
+		Timeout: timeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShareInvitation); ok {
+		return v, err
+	}
+
+	return nil, err
+}
+
+// ResourceShareOwnedBySelfDisassociated waits for a ResourceShare owned by own account to be disassociated
+func ResourceShareOwnedBySelfDisassociated(conn *ram.RAM, arn string, timeout time.Duration) (*ram.ResourceShare, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareAssociationStatusAssociated},
+		Target:  []string{},
+		Refresh: ResourceShareOwnerSelfStatus(conn, arn),
+		Timeout: timeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShare); ok {
+		return v, err
+	}
+
+	return nil, err
+}
+
+// ResourceShareOwnedBySelfActive waits for a ResourceShare owned by own account to return ACTIVE
+func ResourceShareOwnedBySelfActive(conn *ram.RAM, arn string, timeout time.Duration) (*ram.ResourceShare, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareStatusPending},
+		Target:  []string{ram.ResourceShareStatusActive},
+		Refresh: ResourceShareOwnerSelfStatus(conn, arn),
+		Timeout: timeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShare); ok {
+		return v, err
+	}
+
+	return nil, err
+}
+
+// ResourceShareOwnedBySelfDeleted waits for a ResourceShare owned by own account to return DELETED
+func ResourceShareOwnedBySelfDeleted(conn *ram.RAM, arn string, timeout time.Duration) (*ram.ResourceShare, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareStatusDeleting},
+		Target:  []string{ram.ResourceShareStatusDeleted},
+		Refresh: ResourceShareOwnerSelfStatus(conn, arn),
+		Timeout: timeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShare); ok {
+		return v, err
+	}
+
+	return nil, err
+}

--- a/aws/internal/service/ram/waiter/waiter.go
+++ b/aws/internal/service/ram/waiter/waiter.go
@@ -7,6 +7,11 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 )
 
+const (
+	PrincipalAssociationTimeout    = 3 * time.Minute
+	PrincipalDisassociationTimeout = 3 * time.Minute
+)
+
 // ResourceShareInvitationAccepted waits for a ResourceShareInvitation to return ACCEPTED
 func ResourceShareInvitationAccepted(conn *ram.RAM, arn string, timeout time.Duration) (*ram.ResourceShareInvitation, error) {
 	stateConf := &resource.StateChangeConf{
@@ -73,6 +78,40 @@ func ResourceShareOwnedBySelfDeleted(conn *ram.RAM, arn string, timeout time.Dur
 	outputRaw, err := stateConf.WaitForState()
 
 	if v, ok := outputRaw.(*ram.ResourceShare); ok {
+		return v, err
+	}
+
+	return nil, err
+}
+
+func ResourceSharePrincipalAssociated(conn *ram.RAM, resourceShareARN, principal string) (*ram.ResourceShareAssociation, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareAssociationStatusAssociating, PrincipalAssociationStatusNotFound},
+		Target:  []string{ram.ResourceShareAssociationStatusAssociated},
+		Refresh: ResourceSharePrincipalAssociationStatus(conn, resourceShareARN, principal),
+		Timeout: PrincipalAssociationTimeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShareAssociation); ok {
+		return v, err
+	}
+
+	return nil, err
+}
+
+func ResourceSharePrincipalDisassociated(conn *ram.RAM, resourceShareARN, principal string) (*ram.ResourceShareAssociation, error) {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{ram.ResourceShareAssociationStatusAssociated, ram.ResourceShareAssociationStatusDisassociating},
+		Target:  []string{ram.ResourceShareAssociationStatusDisassociated, PrincipalAssociationStatusNotFound},
+		Refresh: ResourceSharePrincipalAssociationStatus(conn, resourceShareARN, principal),
+		Timeout: PrincipalDisassociationTimeout,
+	}
+
+	outputRaw, err := stateConf.WaitForState()
+
+	if v, ok := outputRaw.(*ram.ResourceShareAssociation); ok {
 		return v, err
 	}
 

--- a/aws/resource_aws_ram_principal_association.go
+++ b/aws/resource_aws_ram_principal_association.go
@@ -5,13 +5,15 @@ import (
 	"log"
 	"regexp"
 	"strings"
-	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ram"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/finder"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/waiter"
 )
 
 func resourceAwsRamPrincipalAssociation() *schema.Resource {
@@ -60,7 +62,7 @@ func resourceAwsRamPrincipalAssociationCreate(d *schema.ResourceData, meta inter
 	log.Println("[DEBUG] Create RAM principal association request:", request)
 	_, err := conn.AssociateResourceShare(request)
 	if err != nil {
-		return fmt.Errorf("Error associating principal with RAM resource share: %s", err)
+		return fmt.Errorf("error associating principal with RAM resource share: %w", err)
 	}
 
 	d.SetId(fmt.Sprintf("%s,%s", resourceShareArn, principal))
@@ -70,8 +72,8 @@ func resourceAwsRamPrincipalAssociationCreate(d *schema.ResourceData, meta inter
 		return resourceAwsRamPrincipalAssociationRead(d, meta)
 	}
 
-	if err := waitForRamResourceSharePrincipalAssociation(conn, resourceShareArn, principal); err != nil {
-		return fmt.Errorf("Error waiting for RAM principal association (%s) to become ready: %s", d.Id(), err)
+	if _, err := waiter.ResourceSharePrincipalAssociated(conn, resourceShareArn, principal); err != nil {
+		return fmt.Errorf("error waiting for RAM principal association (%s) to become ready: %w", d.Id(), err)
 	}
 
 	return resourceAwsRamPrincipalAssociationRead(d, meta)
@@ -82,25 +84,36 @@ func resourceAwsRamPrincipalAssociationRead(d *schema.ResourceData, meta interfa
 
 	resourceShareArn, principal, err := resourceAwsRamPrincipalAssociationParseId(d.Id())
 	if err != nil {
-		return err
+		return fmt.Errorf("error reading RAM Principal Association, parsing ID (%s): %w", d.Id(), err)
 	}
 
-	resourceShareAssociation, err := getRamResourceSharePrincipalAssociation(conn, resourceShareArn, principal)
+	var association *ram.ResourceShareAssociation
+
+	if ok, _ := regexp.MatchString(`^\d{12}$`, principal); ok {
+		// AWS Account ID Principals need to be accepted to become ASSOCIATED
+		association, err = finder.ResourceSharePrincipalAssociationByShareARNPrincipal(conn, resourceShareArn, principal)
+	} else {
+		association, err = waiter.ResourceSharePrincipalAssociated(conn, resourceShareArn, principal)
+	}
+
+	if !d.IsNewResource() && (tfawserr.ErrCodeEquals(err, ram.ErrCodeResourceArnNotFoundException) || tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException)) {
+		log.Printf("[WARN] No RAM resource share principal association with ARN (%s) found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
 
 	if err != nil {
 		return fmt.Errorf("error reading RAM Resource Share (%s) Principal Association (%s): %s", resourceShareArn, principal, err)
 	}
 
-	if resourceShareAssociation == nil {
-		log.Printf("[WARN] RAM Resource Share (%s) Principal Association (%s) not found, removing from state", resourceShareArn, principal)
+	if association == nil || aws.StringValue(association.Status) == ram.ResourceShareAssociationStatusDisassociated {
+		log.Printf("[WARN] RAM resource share principal association with ARN (%s) found, but empty or disassociated - removing from state", d.Id())
 		d.SetId("")
 		return nil
 	}
 
-	if aws.StringValue(resourceShareAssociation.Status) != ram.ResourceShareAssociationStatusAssociated && aws.StringValue(resourceShareAssociation.Status) != ram.ResourceShareAssociationStatusAssociating {
-		log.Printf("[WARN] RAM Resource Share (%s) Principal Association (%s) not associating or associated, removing from state", resourceShareArn, principal)
-		d.SetId("")
-		return nil
+	if aws.StringValue(association.Status) != ram.ResourceShareAssociationStatusAssociated && aws.StringValue(association.Status) != ram.ResourceShareAssociationStatusAssociating {
+		return fmt.Errorf("error reading RAM Resource Share (%s) Principal Association (%s), status not associating or associated: %s", resourceShareArn, principal, aws.StringValue(association.Status))
 	}
 
 	d.Set("resource_share_arn", resourceShareArn)
@@ -125,7 +138,7 @@ func resourceAwsRamPrincipalAssociationDelete(d *schema.ResourceData, meta inter
 	log.Println("[DEBUG] Delete RAM principal association request:", request)
 	_, err = conn.DisassociateResourceShare(request)
 
-	if isAWSErr(err, ram.ErrCodeUnknownResourceException, "") {
+	if tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException) {
 		return nil
 	}
 
@@ -133,7 +146,7 @@ func resourceAwsRamPrincipalAssociationDelete(d *schema.ResourceData, meta inter
 		return fmt.Errorf("error disassociating RAM Resource Share (%s) Principal Association (%s): %s", resourceShareArn, principal, err)
 	}
 
-	if err := waitForRamResourceSharePrincipalDisassociation(conn, resourceShareArn, principal); err != nil {
+	if _, err := waiter.ResourceSharePrincipalDisassociated(conn, resourceShareArn, principal); err != nil {
 		return fmt.Errorf("error waiting for RAM Resource Share (%s) Principal Association (%s) disassociation: %s", resourceShareArn, principal, err)
 	}
 
@@ -149,75 +162,4 @@ func resourceAwsRamPrincipalAssociationParseId(id string) (string, string, error
 	}
 
 	return parts[0], parts[1], nil
-}
-
-func getRamResourceSharePrincipalAssociation(conn *ram.RAM, resourceShareARN, principal string) (*ram.ResourceShareAssociation, error) {
-	input := &ram.GetResourceShareAssociationsInput{
-		AssociationType:   aws.String(ram.ResourceShareAssociationTypePrincipal),
-		Principal:         aws.String(principal),
-		ResourceShareArns: aws.StringSlice([]string{resourceShareARN}),
-	}
-
-	output, err := conn.GetResourceShareAssociations(input)
-
-	if isAWSErr(err, ram.ErrCodeUnknownResourceException, "") {
-		return nil, nil
-	}
-
-	if err != nil {
-		return nil, err
-	}
-
-	if output == nil || len(output.ResourceShareAssociations) == 0 || output.ResourceShareAssociations[0] == nil {
-		return nil, nil
-	}
-
-	return output.ResourceShareAssociations[0], nil
-}
-
-func resourceAwsRamPrincipalAssociationStateRefreshFunc(conn *ram.RAM, resourceShareArn, principal string) resource.StateRefreshFunc {
-	return func() (interface{}, string, error) {
-		association, err := getRamResourceSharePrincipalAssociation(conn, resourceShareArn, principal)
-
-		if err != nil {
-			return nil, ram.ResourceShareAssociationStatusFailed, err
-		}
-
-		if association == nil {
-			return nil, ram.ResourceShareAssociationStatusDisassociated, nil
-		}
-
-		if aws.StringValue(association.Status) == ram.ResourceShareAssociationStatusFailed {
-			extendedErr := fmt.Errorf("association status message: %s", aws.StringValue(association.StatusMessage))
-			return association, aws.StringValue(association.Status), extendedErr
-		}
-
-		return association, aws.StringValue(association.Status), nil
-	}
-}
-
-func waitForRamResourceSharePrincipalAssociation(conn *ram.RAM, resourceShareARN, principal string) error {
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareAssociationStatusAssociating},
-		Target:  []string{ram.ResourceShareAssociationStatusAssociated},
-		Refresh: resourceAwsRamPrincipalAssociationStateRefreshFunc(conn, resourceShareARN, principal),
-		Timeout: 5 * time.Minute,
-	}
-
-	_, err := stateConf.WaitForState()
-
-	return err
-}
-
-func waitForRamResourceSharePrincipalDisassociation(conn *ram.RAM, resourceShareARN, principal string) error {
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareAssociationStatusAssociated, ram.ResourceShareAssociationStatusDisassociating},
-		Target:  []string{ram.ResourceShareAssociationStatusDisassociated},
-		Refresh: resourceAwsRamPrincipalAssociationStateRefreshFunc(conn, resourceShareARN, principal),
-		Timeout: 5 * time.Minute,
-	}
-
-	_, err := stateConf.WaitForState()
-
-	return err
 }

--- a/aws/resource_aws_ram_principal_association_test.go
+++ b/aws/resource_aws_ram_principal_association_test.go
@@ -2,6 +2,7 @@ package aws
 
 import (
 	"fmt"
+	"regexp"
 	"testing"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -9,10 +10,12 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/finder"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/waiter"
 )
 
 func TestAccAwsRamPrincipalAssociation_basic(t *testing.T) {
-	var resourceShareAssociation1 ram.ResourceShareAssociation
+	var association ram.ResourceShareAssociation
 	resourceName := "aws_ram_principal_association.test"
 	rName := acctest.RandomWithPrefix("tf-acc-test")
 
@@ -25,7 +28,7 @@ func TestAccAwsRamPrincipalAssociation_basic(t *testing.T) {
 			{
 				Config: testAccAwsRamPrincipalAssociationConfig(rName),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckAwsRamPrincipalAssociationExists(resourceName, &resourceShareAssociation1),
+					testAccCheckAwsRamPrincipalAssociationExists(resourceName, &association),
 				),
 			},
 			{
@@ -38,7 +41,7 @@ func TestAccAwsRamPrincipalAssociation_basic(t *testing.T) {
 }
 
 func TestAccAwsRamPrincipalAssociation_disappears(t *testing.T) {
-	var resourceShareAssociation1 ram.ResourceShareAssociation
+	var association ram.ResourceShareAssociation
 	resourceName := "aws_ram_principal_association.test"
 	rName := acctest.RandomWithPrefix("tf-acc-test")
 
@@ -46,36 +49,18 @@ func TestAccAwsRamPrincipalAssociation_disappears(t *testing.T) {
 		PreCheck:     func() { testAccPreCheck(t) },
 		ErrorCheck:   testAccErrorCheck(t, ram.EndpointsID),
 		Providers:    testAccProviders,
-		CheckDestroy: testAccCheckAwsRamResourceAssociationDestroy,
+		CheckDestroy: testAccCheckAwsRamPrincipalAssociationDestroy,
 		Steps: []resource.TestStep{
 			{
 				Config: testAccAwsRamPrincipalAssociationConfig(rName),
 				Check: resource.ComposeTestCheckFunc(
-					testAccCheckAwsRamPrincipalAssociationExists(resourceName, &resourceShareAssociation1),
-					testAccCheckAwsRamPrincipalAssociationDisappears(&resourceShareAssociation1),
+					testAccCheckAwsRamPrincipalAssociationExists(resourceName, &association),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsRamPrincipalAssociation(), resourceName),
 				),
 				ExpectNonEmptyPlan: true,
 			},
 		},
 	})
-}
-
-func testAccCheckAwsRamPrincipalAssociationDisappears(resourceShareAssociation *ram.ResourceShareAssociation) resource.TestCheckFunc {
-	return func(s *terraform.State) error {
-		conn := testAccProvider.Meta().(*AWSClient).ramconn
-
-		input := &ram.DisassociateResourceShareInput{
-			Principals:       []*string{resourceShareAssociation.AssociatedEntity},
-			ResourceShareArn: resourceShareAssociation.ResourceShareArn,
-		}
-
-		_, err := conn.DisassociateResourceShare(input)
-		if err != nil {
-			return err
-		}
-
-		return waitForRamResourceSharePrincipalDisassociation(conn, aws.StringValue(resourceShareAssociation.ResourceShareArn), aws.StringValue(resourceShareAssociation.AssociatedEntity))
-	}
 }
 
 func testAccCheckAwsRamPrincipalAssociationExists(resourceName string, resourceShare *ram.ResourceShareAssociation) resource.TestCheckFunc {
@@ -94,24 +79,31 @@ func testAccCheckAwsRamPrincipalAssociationExists(resourceName string, resourceS
 		resourceShareARN, principal, err := resourceAwsRamPrincipalAssociationParseId(rs.Primary.ID)
 
 		if err != nil {
-			return err
+			return fmt.Errorf("error parsing ID (%s): %w", rs.Primary.ID, err)
 		}
 
-		resourceShareAssociation, err := getRamResourceSharePrincipalAssociation(conn, resourceShareARN, principal)
+		var association *ram.ResourceShareAssociation
+
+		if ok, _ := regexp.MatchString(`^\d{12}$`, principal); ok {
+			// AWS Account ID Principals need to be accepted to become ASSOCIATED
+			association, err = finder.ResourceSharePrincipalAssociationByShareARNPrincipal(conn, resourceShareARN, principal)
+		} else {
+			association, err = waiter.ResourceSharePrincipalAssociated(conn, resourceShareARN, principal)
+		}
 
 		if err != nil {
 			return fmt.Errorf("error reading RAM Resource Share (%s) Principal Association (%s): %s", resourceShareARN, principal, err)
 		}
 
-		if resourceShareAssociation == nil {
+		if association == nil {
 			return fmt.Errorf("RAM Resource Share (%s) Principal Association (%s) not found", resourceShareARN, principal)
 		}
 
-		if aws.StringValue(resourceShareAssociation.Status) != ram.ResourceShareAssociationStatusAssociated && aws.StringValue(resourceShareAssociation.Status) != ram.ResourceShareAssociationStatusAssociating {
-			return fmt.Errorf("RAM Resource Share (%s) Principal Association (%s) not associating or associated: %s", resourceShareARN, principal, aws.StringValue(resourceShareAssociation.Status))
+		if aws.StringValue(association.Status) != ram.ResourceShareAssociationStatusAssociated && aws.StringValue(association.Status) != ram.ResourceShareAssociationStatusAssociating {
+			return fmt.Errorf("RAM Resource Share (%s) Principal Association (%s) status not associating or associated: %s", resourceShareARN, principal, aws.StringValue(association.Status))
 		}
 
-		*resourceShare = *resourceShareAssociation
+		*resourceShare = *association
 
 		return nil
 	}
@@ -131,14 +123,14 @@ func testAccCheckAwsRamPrincipalAssociationDestroy(s *terraform.State) error {
 			return err
 		}
 
-		resourceShareAssociation, err := getRamResourceSharePrincipalAssociation(conn, resourceShareARN, principal)
+		association, err := waiter.ResourceSharePrincipalDisassociated(conn, resourceShareARN, principal)
 
 		if err != nil {
 			return err
 		}
 
-		if resourceShareAssociation != nil && aws.StringValue(resourceShareAssociation.Status) != ram.ResourceShareAssociationStatusDisassociated {
-			return fmt.Errorf("RAM Resource Share (%s) Principal Association (%s) not disassociated: %s", resourceShareARN, principal, aws.StringValue(resourceShareAssociation.Status))
+		if association != nil && aws.StringValue(association.Status) != ram.ResourceShareAssociationStatusDisassociated {
+			return fmt.Errorf("RAM Resource Share (%s) Principal Association (%s) not disassociated: %s", resourceShareARN, principal, aws.StringValue(association.Status))
 		}
 	}
 

--- a/aws/resource_aws_ram_resource_share.go
+++ b/aws/resource_aws_ram_resource_share.go
@@ -7,9 +7,9 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ram"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/waiter"
 )
 
 func resourceAwsRamResourceShare() *schema.Resource {
@@ -75,14 +75,7 @@ func resourceAwsRamResourceShareCreate(d *schema.ResourceData, meta interface{})
 
 	d.SetId(aws.StringValue(createResp.ResourceShare.ResourceShareArn))
 
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareStatusPending},
-		Target:  []string{ram.ResourceShareStatusActive},
-		Refresh: resourceAwsRamResourceShareStateRefreshFunc(conn, d.Id()),
-		Timeout: d.Timeout(schema.TimeoutCreate),
-	}
-
-	_, err = stateConf.WaitForState()
+	_, err = waiter.ResourceShareOwnedBySelfActive(conn, d.Id(), d.Timeout(schema.TimeoutCreate))
 	if err != nil {
 		return fmt.Errorf("Error waiting for RAM resource share (%s) to become ready: %s", d.Id(), err)
 	}
@@ -191,40 +184,11 @@ func resourceAwsRamResourceShareDelete(d *schema.ResourceData, meta interface{})
 		return fmt.Errorf("Error deleting RAM resource share %s: %s", d.Id(), err)
 	}
 
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareStatusDeleting},
-		Target:  []string{ram.ResourceShareStatusDeleted},
-		Refresh: resourceAwsRamResourceShareStateRefreshFunc(conn, d.Id()),
-		Timeout: d.Timeout(schema.TimeoutDelete),
-	}
+	_, err = waiter.ResourceShareOwnedBySelfDeleted(conn, d.Id(), d.Timeout(schema.TimeoutDelete))
 
-	_, err = stateConf.WaitForState()
 	if err != nil {
 		return fmt.Errorf("Error waiting for RAM resource share (%s) to become ready: %s", d.Id(), err)
 	}
 
 	return nil
-}
-
-func resourceAwsRamResourceShareStateRefreshFunc(conn *ram.RAM, resourceShareArn string) resource.StateRefreshFunc {
-	return func() (interface{}, string, error) {
-		request := &ram.GetResourceSharesInput{
-			ResourceShareArns: []*string{aws.String(resourceShareArn)},
-			ResourceOwner:     aws.String(ram.ResourceOwnerSelf),
-		}
-
-		output, err := conn.GetResourceShares(request)
-
-		if err != nil {
-			return nil, ram.ResourceShareStatusFailed, err
-		}
-
-		if len(output.ResourceShares) == 0 {
-			return nil, ram.ResourceShareStatusDeleted, nil
-		}
-
-		resourceShare := output.ResourceShares[0]
-
-		return resourceShare, aws.StringValue(resourceShare.Status), nil
-	}
 }

--- a/aws/resource_aws_ram_resource_share_accepter.go
+++ b/aws/resource_aws_ram_resource_share_accepter.go
@@ -8,6 +8,7 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ram"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/finder"
@@ -129,8 +130,8 @@ func resourceAwsRamResourceShareAccepterRead(d *schema.ResourceData, meta interf
 
 	invitation, err := finder.ResourceShareInvitationByResourceShareArnAndStatus(conn, d.Id(), ram.ResourceShareInvitationStatusAccepted)
 
-	if err != nil {
-		return fmt.Errorf("Error retrieving invitation for resource share %s: %s", d.Id(), err)
+	if err != nil && !tfawserr.ErrCodeEquals(err, ram.ErrCodeResourceShareInvitationArnNotFoundException) {
+		return fmt.Errorf("error retrieving invitation for resource share %s: %w", d.Id(), err)
 	}
 
 	if invitation != nil {
@@ -142,14 +143,18 @@ func resourceAwsRamResourceShareAccepterRead(d *schema.ResourceData, meta interf
 
 	resourceShare, err := finder.ResourceShareOwnerOtherAccountsByArn(conn, d.Id())
 
+	if !d.IsNewResource() && (tfawserr.ErrCodeEquals(err, ram.ErrCodeResourceArnNotFoundException) || tfawserr.ErrCodeEquals(err, ram.ErrCodeUnknownResourceException)) {
+		log.Printf("[WARN] No RAM resource share with ARN (%s) found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
 	if err != nil {
 		return fmt.Errorf("error retrieving resource share: %w", err)
 	}
 
 	if resourceShare == nil {
-		log.Printf("[WARN] No RAM resource share with ARN (%s) found, removing from state", d.Id())
-		d.SetId("")
-		return nil
+		return fmt.Errorf("error getting resource share (%s): empty result", d.Id())
 	}
 
 	d.Set("status", resourceShare.Status)

--- a/aws/resource_aws_ram_resource_share_accepter.go
+++ b/aws/resource_aws_ram_resource_share_accepter.go
@@ -10,6 +10,8 @@ import (
 	"github.com/aws/aws-sdk-go/service/ram"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/finder"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/ram/waiter"
 )
 
 func resourceAwsRamResourceShareAccepter() *schema.Resource {
@@ -81,7 +83,7 @@ func resourceAwsRamResourceShareAccepterCreate(d *schema.ResourceData, meta inte
 
 	shareARN := d.Get("share_arn").(string)
 
-	invitation, err := resourceAwsRamResourceShareGetInvitation(conn, shareARN, ram.ResourceShareInvitationStatusPending)
+	invitation, err := finder.ResourceShareInvitationByResourceShareArnAndStatus(conn, shareARN, ram.ResourceShareInvitationStatusPending)
 
 	if err != nil {
 		return err
@@ -108,16 +110,11 @@ func resourceAwsRamResourceShareAccepterCreate(d *schema.ResourceData, meta inte
 
 	d.SetId(shareARN)
 
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareInvitationStatusPending},
-		Target:  []string{ram.ResourceShareInvitationStatusAccepted},
-		Refresh: resourceAwsRamResourceShareAccepterStateRefreshFunc(
-			conn,
-			aws.StringValue(output.ResourceShareInvitation.ResourceShareInvitationArn)),
-		Timeout: d.Timeout(schema.TimeoutCreate),
-	}
-
-	_, err = stateConf.WaitForState()
+	_, err = waiter.ResourceShareInvitationAccepted(
+		conn,
+		aws.StringValue(output.ResourceShareInvitation.ResourceShareInvitationArn),
+		d.Timeout(schema.TimeoutCreate),
+	)
 
 	if err != nil {
 		return fmt.Errorf("Error waiting for RAM resource share (%s) state: %s", d.Id(), err)
@@ -130,7 +127,7 @@ func resourceAwsRamResourceShareAccepterRead(d *schema.ResourceData, meta interf
 	accountID := meta.(*AWSClient).accountid
 	conn := meta.(*AWSClient).ramconn
 
-	invitation, err := resourceAwsRamResourceShareGetInvitation(conn, d.Id(), ram.ResourceShareInvitationStatusAccepted)
+	invitation, err := finder.ResourceShareInvitationByResourceShareArnAndStatus(conn, d.Id(), ram.ResourceShareInvitationStatusAccepted)
 
 	if err != nil {
 		return fmt.Errorf("Error retrieving invitation for resource share %s: %s", d.Id(), err)
@@ -143,23 +140,17 @@ func resourceAwsRamResourceShareAccepterRead(d *schema.ResourceData, meta interf
 		d.Set("receiver_account_id", accountID)
 	}
 
-	listResourceSharesInput := &ram.GetResourceSharesInput{
-		ResourceOwner:     aws.String(ram.ResourceOwnerOtherAccounts),
-		ResourceShareArns: aws.StringSlice([]string{d.Id()}),
-	}
+	resourceShare, err := finder.ResourceShareOwnerOtherAccountsByArn(conn, d.Id())
 
-	shares, err := conn.GetResourceShares(listResourceSharesInput)
 	if err != nil {
-		return fmt.Errorf("error retrieving resource shares: %w", err)
+		return fmt.Errorf("error retrieving resource share: %w", err)
 	}
 
-	if len(shares.ResourceShares) != 1 {
+	if resourceShare == nil {
 		log.Printf("[WARN] No RAM resource share with ARN (%s) found, removing from state", d.Id())
 		d.SetId("")
 		return nil
 	}
-
-	resourceShare := shares.ResourceShares[0]
 
 	d.Set("status", resourceShare.Status)
 	d.Set("sender_account_id", resourceShare.OwningAccountId)
@@ -215,72 +206,12 @@ func resourceAwsRamResourceShareAccepterDelete(d *schema.ResourceData, meta inte
 		return fmt.Errorf("Error leaving RAM resource share: %s", err)
 	}
 
-	stateConf := &resource.StateChangeConf{
-		Pending: []string{ram.ResourceShareAssociationStatusAssociated},
-		Target:  []string{ram.ResourceShareAssociationStatusDisassociated},
-		Refresh: resourceAwsRamResourceShareStateRefreshFunc(conn, d.Id()),
-		Timeout: d.Timeout(schema.TimeoutDelete),
-	}
-
-	if _, err := stateConf.WaitForState(); err != nil {
-		if isAWSErr(err, ram.ErrCodeUnknownResourceException, "") {
-			// what we want
-			return nil
-		}
+	_, err = waiter.ResourceShareOwnedBySelfDisassociated(conn, d.Id(), d.Timeout(schema.TimeoutDelete))
+	if err != nil {
 		return fmt.Errorf("Error waiting for RAM resource share (%s) state: %s", d.Id(), err)
 	}
 
 	return nil
-}
-
-func resourceAwsRamResourceShareGetInvitation(conn *ram.RAM, resourceShareARN, status string) (*ram.ResourceShareInvitation, error) {
-	input := &ram.GetResourceShareInvitationsInput{
-		ResourceShareArns: []*string{aws.String(resourceShareARN)},
-	}
-
-	var invitation *ram.ResourceShareInvitation
-	err := conn.GetResourceShareInvitationsPages(input, func(page *ram.GetResourceShareInvitationsOutput, lastPage bool) bool {
-		for _, rsi := range page.ResourceShareInvitations {
-			if aws.StringValue(rsi.Status) == status {
-				invitation = rsi
-				return false
-			}
-		}
-
-		return !lastPage
-	})
-
-	if invitation == nil {
-		return nil, nil
-	}
-
-	if err != nil {
-		return nil, fmt.Errorf("Error reading RAM resource share invitation %s: %s", resourceShareARN, err)
-	}
-
-	return invitation, nil
-}
-
-func resourceAwsRamResourceShareAccepterStateRefreshFunc(conn *ram.RAM, invitationArn string) resource.StateRefreshFunc {
-	return func() (interface{}, string, error) {
-		request := &ram.GetResourceShareInvitationsInput{
-			ResourceShareInvitationArns: []*string{aws.String(invitationArn)},
-		}
-
-		output, err := conn.GetResourceShareInvitations(request)
-
-		if err != nil {
-			return nil, "Unable to get resource share invitations", err
-		}
-
-		if len(output.ResourceShareInvitations) == 0 {
-			return nil, "Resource share invitation not found", nil
-		}
-
-		invitation := output.ResourceShareInvitations[0]
-
-		return invitation, aws.StringValue(invitation.Status), nil
-	}
 }
 
 func resourceAwsRamResourceShareGetIDFromARN(arn string) string {

--- a/aws/resource_aws_ram_resource_share_accepter_test.go
+++ b/aws/resource_aws_ram_resource_share_accepter_test.go
@@ -17,8 +17,7 @@ func TestAccAwsRamResourceShareAccepter_basic(t *testing.T) {
 	var providers []*schema.Provider
 	resourceName := "aws_ram_resource_share_accepter.test"
 	principalAssociationResourceName := "aws_ram_principal_association.test"
-
-	shareName := fmt.Sprintf("tf-%s", acctest.RandString(10))
+	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck: func() {
@@ -30,7 +29,7 @@ func TestAccAwsRamResourceShareAccepter_basic(t *testing.T) {
 		CheckDestroy:      testAccCheckAwsRamResourceShareAccepterDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAwsRamResourceShareAccepterBasic(shareName),
+				Config: testAccAwsRamResourceShareAccepterBasic(rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareAccepterExists(resourceName),
 					resource.TestCheckResourceAttrPair(resourceName, "share_arn", principalAssociationResourceName, "resource_share_arn"),
@@ -39,12 +38,12 @@ func TestAccAwsRamResourceShareAccepter_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "status", ram.ResourceShareStatusActive),
 					testAccCheckResourceAttrAccountID(resourceName, "receiver_account_id"),
 					resource.TestMatchResourceAttr(resourceName, "sender_account_id", regexp.MustCompile(`\d{12}`)),
-					resource.TestCheckResourceAttr(resourceName, "share_name", shareName),
+					resource.TestCheckResourceAttr(resourceName, "share_name", rName),
 					resource.TestCheckResourceAttr(resourceName, "resources.%", "0"),
 				),
 			},
 			{
-				Config:            testAccAwsRamResourceShareAccepterBasic(shareName),
+				Config:            testAccAwsRamResourceShareAccepterBasic(rName),
 				ResourceName:      resourceName,
 				ImportState:       true,
 				ImportStateVerify: true,
@@ -132,7 +131,7 @@ func testAccCheckAwsRamResourceShareAccepterExists(name string) resource.TestChe
 	}
 }
 
-func testAccAwsRamResourceShareAccepterBasic(shareName string) string {
+func testAccAwsRamResourceShareAccepterBasic(rName string) string {
 	return testAccAlternateAccountProviderConfig() + fmt.Sprintf(`
 resource "aws_ram_resource_share_accepter" "test" {
   share_arn = aws_ram_principal_association.test.resource_share_arn
@@ -157,5 +156,5 @@ resource "aws_ram_principal_association" "test" {
 }
 
 data "aws_caller_identity" "receiver" {}
-`, shareName)
+`, rName)
 }

--- a/aws/resource_aws_ram_resource_share_accepter_test.go
+++ b/aws/resource_aws_ram_resource_share_accepter_test.go
@@ -53,6 +53,32 @@ func TestAccAwsRamResourceShareAccepter_basic(t *testing.T) {
 	})
 }
 
+func TestAccAwsRamResourceShareAccepter_disappears(t *testing.T) {
+	var providers []*schema.Provider
+	resourceName := "aws_ram_resource_share_accepter.test"
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			testAccAlternateAccountPreCheck(t)
+		},
+		ErrorCheck:        testAccErrorCheck(t, ram.EndpointsID),
+		ProviderFactories: testAccProviderFactoriesAlternate(&providers),
+		CheckDestroy:      testAccCheckAwsRamResourceShareAccepterDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAwsRamResourceShareAccepterBasic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsRamResourceShareAccepterExists(resourceName),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsRamResourceShareAccepter(), resourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
 func testAccCheckAwsRamResourceShareAccepterDestroy(s *terraform.State) error {
 	conn := testAccProvider.Meta().(*AWSClient).ramconn
 

--- a/aws/resource_aws_ram_resource_share_test.go
+++ b/aws/resource_aws_ram_resource_share_test.go
@@ -14,8 +14,8 @@ import (
 
 func TestAccAwsRamResourceShare_basic(t *testing.T) {
 	var resourceShare ram.ResourceShare
-	resourceName := "aws_ram_resource_share.example"
-	shareName := fmt.Sprintf("tf-%s", acctest.RandString(10))
+	resourceName := "aws_ram_resource_share.test"
+	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -24,12 +24,12 @@ func TestAccAwsRamResourceShare_basic(t *testing.T) {
 		CheckDestroy: testAccCheckAwsRamResourceShareDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAwsRamResourceShareConfigName(shareName),
+				Config: testAccAwsRamResourceShareConfigName(rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare),
 					testAccMatchResourceAttrRegionalARN(resourceName, "arn", "ram", regexp.MustCompile(`resource-share/.+`)),
 					resource.TestCheckResourceAttr(resourceName, "allow_external_principals", "false"),
-					resource.TestCheckResourceAttr(resourceName, "name", shareName),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 				),
 			},
@@ -44,8 +44,8 @@ func TestAccAwsRamResourceShare_basic(t *testing.T) {
 
 func TestAccAwsRamResourceShare_AllowExternalPrincipals(t *testing.T) {
 	var resourceShare1, resourceShare2 ram.ResourceShare
-	resourceName := "aws_ram_resource_share.example"
-	shareName := fmt.Sprintf("tf-%s", acctest.RandString(10))
+	resourceName := "aws_ram_resource_share.test"
+	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -54,7 +54,7 @@ func TestAccAwsRamResourceShare_AllowExternalPrincipals(t *testing.T) {
 		CheckDestroy: testAccCheckAwsRamResourceShareDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAwsRamResourceShareConfigAllowExternalPrincipals(shareName, false),
+				Config: testAccAwsRamResourceShareConfigAllowExternalPrincipals(rName, false),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare1),
 					resource.TestCheckResourceAttr(resourceName, "allow_external_principals", "false"),
@@ -66,7 +66,7 @@ func TestAccAwsRamResourceShare_AllowExternalPrincipals(t *testing.T) {
 				ImportStateVerify: true,
 			},
 			{
-				Config: testAccAwsRamResourceShareConfigAllowExternalPrincipals(shareName, true),
+				Config: testAccAwsRamResourceShareConfigAllowExternalPrincipals(rName, true),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare2),
 					resource.TestCheckResourceAttr(resourceName, "allow_external_principals", "true"),
@@ -78,9 +78,9 @@ func TestAccAwsRamResourceShare_AllowExternalPrincipals(t *testing.T) {
 
 func TestAccAwsRamResourceShare_Name(t *testing.T) {
 	var resourceShare1, resourceShare2 ram.ResourceShare
-	resourceName := "aws_ram_resource_share.example"
-	shareName1 := fmt.Sprintf("tf-%s", acctest.RandString(10))
-	shareName2 := fmt.Sprintf("tf-%s", acctest.RandString(10))
+	resourceName := "aws_ram_resource_share.test"
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+	rName2 := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -89,10 +89,10 @@ func TestAccAwsRamResourceShare_Name(t *testing.T) {
 		CheckDestroy: testAccCheckAwsRamResourceShareDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAwsRamResourceShareConfigName(shareName1),
+				Config: testAccAwsRamResourceShareConfigName(rName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare1),
-					resource.TestCheckResourceAttr(resourceName, "name", shareName1),
+					resource.TestCheckResourceAttr(resourceName, "name", rName),
 				),
 			},
 			{
@@ -101,10 +101,10 @@ func TestAccAwsRamResourceShare_Name(t *testing.T) {
 				ImportStateVerify: true,
 			},
 			{
-				Config: testAccAwsRamResourceShareConfigName(shareName2),
+				Config: testAccAwsRamResourceShareConfigName(rName2),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare2),
-					resource.TestCheckResourceAttr(resourceName, "name", shareName2),
+					resource.TestCheckResourceAttr(resourceName, "name", rName2),
 				),
 			},
 		},
@@ -113,8 +113,8 @@ func TestAccAwsRamResourceShare_Name(t *testing.T) {
 
 func TestAccAwsRamResourceShare_Tags(t *testing.T) {
 	var resourceShare1, resourceShare2, resourceShare3 ram.ResourceShare
-	resourceName := "aws_ram_resource_share.example"
-	shareName := fmt.Sprintf("tf-%s", acctest.RandString(10))
+	resourceName := "aws_ram_resource_share.test"
+	rName := acctest.RandomWithPrefix("tf-acc-test")
 
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -123,7 +123,7 @@ func TestAccAwsRamResourceShare_Tags(t *testing.T) {
 		CheckDestroy: testAccCheckAwsRamResourceShareDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAwsRamResourceShareConfigTags1(shareName, "key1", "value1"),
+				Config: testAccAwsRamResourceShareConfigTags1(rName, "key1", "value1"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare1),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
@@ -136,7 +136,7 @@ func TestAccAwsRamResourceShare_Tags(t *testing.T) {
 				ImportStateVerify: true,
 			},
 			{
-				Config: testAccAwsRamResourceShareConfigTags2(shareName, "key1", "value1updated", "key2", "value2"),
+				Config: testAccAwsRamResourceShareConfigTags2(rName, "key1", "value1updated", "key2", "value2"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare2),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "2"),
@@ -145,7 +145,7 @@ func TestAccAwsRamResourceShare_Tags(t *testing.T) {
 				),
 			},
 			{
-				Config: testAccAwsRamResourceShareConfigTags1(shareName, "key2", "value2"),
+				Config: testAccAwsRamResourceShareConfigTags1(rName, "key2", "value2"),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAwsRamResourceShareExists(resourceName, &resourceShare3),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
@@ -222,44 +222,44 @@ func testAccCheckAwsRamResourceShareDestroy(s *terraform.State) error {
 	return nil
 }
 
-func testAccAwsRamResourceShareConfigAllowExternalPrincipals(shareName string, allowExternalPrincipals bool) string {
+func testAccAwsRamResourceShareConfigAllowExternalPrincipals(rName string, allowExternalPrincipals bool) string {
 	return fmt.Sprintf(`
-resource "aws_ram_resource_share" "example" {
-  allow_external_principals = %t
-  name                      = %q
+resource "aws_ram_resource_share" "test" {
+  allow_external_principals = %[1]t
+  name                      = %[2]q
 }
-`, allowExternalPrincipals, shareName)
+`, allowExternalPrincipals, rName)
 }
 
-func testAccAwsRamResourceShareConfigName(shareName string) string {
+func testAccAwsRamResourceShareConfigName(rName string) string {
 	return fmt.Sprintf(`
-resource "aws_ram_resource_share" "example" {
-  name = %q
+resource "aws_ram_resource_share" "test" {
+  name = %[1]q
 }
-`, shareName)
+`, rName)
 }
 
-func testAccAwsRamResourceShareConfigTags1(shareName, tagKey1, tagValue1 string) string {
+func testAccAwsRamResourceShareConfigTags1(rName, tagKey1, tagValue1 string) string {
 	return fmt.Sprintf(`
-resource "aws_ram_resource_share" "example" {
-  name = %q
+resource "aws_ram_resource_share" "test" {
+  name = %[1]q
 
   tags = {
-    %q = %q
+    %[2]q = %[3]q
   }
 }
-`, shareName, tagKey1, tagValue1)
+`, rName, tagKey1, tagValue1)
 }
 
-func testAccAwsRamResourceShareConfigTags2(shareName, tagKey1, tagValue1, tagKey2, tagValue2 string) string {
+func testAccAwsRamResourceShareConfigTags2(rName, tagKey1, tagValue1, tagKey2, tagValue2 string) string {
 	return fmt.Sprintf(`
-resource "aws_ram_resource_share" "example" {
-  name = %q
+resource "aws_ram_resource_share" "test" {
+  name = %[1]q
 
   tags = {
-    %q = %q
-    %q = %q
+    %[2]q = %[3]q
+    %[4]q = %[5]q
   }
 }
-`, shareName, tagKey1, tagValue1, tagKey2, tagValue2)
+`, rName, tagKey1, tagValue1, tagKey2, tagValue2)
 }

